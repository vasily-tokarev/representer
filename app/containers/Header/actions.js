/*
 *
 * Post actions
 *
 */

import {
  HELP_TOGGLE,
} from './constants';

export function toggleHelp() {
  return {
    type: HELP_TOGGLE,
  };
}

