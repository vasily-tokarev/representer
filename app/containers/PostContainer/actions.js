/*
 *
 * Post actions
 *
 */

import {
  LOAD_POST_SUCCESS,
  LOAD_POST,
  UNMOUNT_POST,
  WS_PAYLOAD,
} from './constants';

export function loadPost(name) {
  return {
    type: LOAD_POST,
    name,
  };
}

export function unmountPost() {
  return {
    type: UNMOUNT_POST,
  };
}

export function postLoaded(post) {
  return {
    type: LOAD_POST_SUCCESS,
    post,
  };
}

export function WSPayload(data) {
  return {
    type: WS_PAYLOAD,
    data,
  };
}
