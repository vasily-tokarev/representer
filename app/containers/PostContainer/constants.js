/*
 *
 * Post constants
 *
 */

export const LOAD_POST_SUCCESS = 'LOAD_POST_SUCCESS';
export const LOAD_POST = 'LOAD_POST';
export const UNMOUNT_POST = 'UNMOUNT_POST';

