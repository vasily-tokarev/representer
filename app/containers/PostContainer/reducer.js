/*
 *
 * Post reducer
 *
 */

import { fromJS } from 'immutable';
import {
  LOAD_POST,
  LOAD_POST_SUCCESS,
  UNMOUNT_POST,
} from './constants';

const initialState = fromJS({});

function postReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_POST:
      return state
        .set('name', action.name);
    case LOAD_POST_SUCCESS:
      return state
        .set('data', action.post);
    case UNMOUNT_POST:
      return state
        .set('data', { id: 0, text: '', name: '', title: 'Loading' });
    default:
      return state;
  }
}

export default postReducer;
