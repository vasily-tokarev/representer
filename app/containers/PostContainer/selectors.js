import { createSelector } from 'reselect';

/**
 * Direct selector to the post state domain
 */
const selectPostDomain = (state) => state.get('post');

/**
 * Other specific selectors
 */


/**
 * Default selector used by Post
 */

const makeSelectPostName = () => createSelector(
  selectPostDomain,
  (substate) => substate.get('name'),
);

const makeSelectPost = () => createSelector(
  selectPostDomain,
  (substate) => substate.toJS(),
);

// Need this?
export default makeSelectPost;
export {
  makeSelectPost,
  selectPostDomain,
  makeSelectPostName,
};
