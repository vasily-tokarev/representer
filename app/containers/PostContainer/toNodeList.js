import React from 'react';
import shortid from 'shortid';

// TODO: Implement `re.exec` version.
// https://stackoverflow.com/questions/6323417/how-do-i-retrieve-all-matches-for-a-regular-expression-in-javascript

const element = (m, fn) => (
  {
    type: 'jsx',
    match: m.map(I),
    jsx: fn,
    index: m.index,
    length: m[0].length,
  }
);

const url = (m) => <a key={id()} href={m[1]}>{m[2]}</a>;
const bold = (m) => <strong key={id()}>{m[1]}</strong>;
const italic = (m) => <i key={id()}>{m[1]}</i>;
const h1 = (m) => <strong key={id()}>{m[1]}<br/></strong>;
const newline = () => <br key={id()}/>;

const id = shortid.generate;

const matchers = [
  [italic, /_(.*?)_/],
  [bold, /\*\*(.*?)\*\*/],
  [url, /\[(.*)]\((.*)\)/],
  [h1, /#(.*)\n/],
  [newline, /\n/],
];

const match = (text, m) => text.match(m[1]);
const matched = (text) => (m) => match(text, m) ? element(match(text, m), m[0]) : false;

const I = (identity) => identity;
const byIndexAsc = (a, b) => a.index - b.index;

const jsxs = (text) => matchers.map(matched(text));

const nextText = (text, fstJSX, sndJSX) =>
  fstJSX ? { type: 'text', match: text.substring(0, fstJSX.index) }
    : sndJSX ? { type: 'text', match: text.substring(0, sndJSX.index) }
    : text.length > 0 ? { type: 'text', match: text }
      : false;

const nextJSX = (text, jsx) => jsx && text.substring(0, jsx.index).length === 0 ? jsx : false;

const remainingText = (text, next) => {
  switch (next.type) {
    case 'jsx':
      return text.substring(next.index + next.length, text.length);
    case 'text':
      return text.substring(next.match.length);
    default:
      return '';
  }
};

const toNodeList = (text) => (rendered) => {
  if (text.length === 0) return rendered;

  const sortedJSXs = jsxs(text)
    .filter(I)
    .sort(byIndexAsc);

  const fstJSX = sortedJSXs[0];
  const sndJSX = sortedJSXs[1];

  const next = nextJSX(text, fstJSX) || nextText(text, fstJSX, sndJSX);
  const convertWithRemaining = toNodeList(remainingText(text, next));

  return next.type === 'jsx' ?
    convertWithRemaining(rendered.concat(nextJSX(text, next)))
    :
    convertWithRemaining(rendered.concat(nextText(text, false, fstJSX)));
};

export default toNodeList;
