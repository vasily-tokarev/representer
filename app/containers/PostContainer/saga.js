import { call, put, select, takeLatest } from 'redux-saga/effects';

import request from 'utils/request';
import { postLoaded } from './actions';
import { LOAD_POST } from './constants';
import { makeSelectPostName } from './selectors';

const config = require('../../../config');

export function* getPost() {
  const name = yield select(makeSelectPostName());
  const requestURL = config.env === 'prod' ? `${name}/${name}.json` : `/posts/name/${name}.json`;

  try {
    const post = yield call(request, requestURL);
    yield put(postLoaded(post));
  } catch (err) {
    // yield put(repoLoadingError(err));
  }
}

export default function* postData() {
  yield takeLatest(LOAD_POST, getPost);
}
