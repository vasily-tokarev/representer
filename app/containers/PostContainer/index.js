/**
 *
 * Post
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import Post from 'components/Post';

import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import { loadPost, unmountPost } from './actions';
import makeSelectPost from './selectors';
import reducer from './reducer';
import saga from './saga';
import toNodeList from './toNodeList';

const html = (el) => el.type === 'text' ? el.match : el.jsx(el.match);

export class PostContainer extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  componentWillMount() {
    this.props.onLoad(this.props.match.params.name);
  }

  componentWillUnmount() {
    this.props.onUnmount();
  }

  render() {
    return this.props.post.data && this.props.post.data.text.length > 0
      ? (<Post
        title={this.props.post.data.title}
        name={this.props.post.data.name}
        text={toNodeList(this.props.post.data.text)([]).map(html)}
      />)
      : null;
  }
}

PostContainer.propTypes = {
  match: PropTypes.object,
  onUnmount: PropTypes.func,
  onLoad: PropTypes.func,
  post: PropTypes.object,
  name: PropTypes.string,
};

const mapStateToProps = createStructuredSelector({
  post: makeSelectPost(),
});

export function mapDispatchToProps(dispatch) {
  return {
    onLoad: (name) => {
      dispatch(loadPost(name));
    },
    onUnmount: () => {
      dispatch(unmountPost());
    },
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'post', reducer });
const withSaga = injectSaga({ key: 'post', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(PostContainer);
