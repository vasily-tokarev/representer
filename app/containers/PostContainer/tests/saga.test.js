/**
 * Test  sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { LOAD_POST } from 'containers/PostContainer/constants';
import { put, takeLatest } from 'redux-saga/effects';
import { postLoaded } from 'containers/PostContainer/actions';
import postData, { getPost } from '../saga';

const post = {
  name: 'post',
  text: 'Post text.',
  title: 'Post Title',
};

describe('getPost Saga', () => {
  let getPostGenerator;

  beforeEach(() => {
    getPostGenerator = getPost();

    const selectDescriptor = getPostGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getPostGenerator.next(post).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the getPost action', () => {
    const response = [{
      name: 'first-post',
      text: 'First post text.',
      title: 'First Post Title',
    }, {
      name: 'second-post',
      text: 'Second post text.',
      title: 'Second Post Title',
    }];
    const putDescriptor = getPostGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(postLoaded(response, response)));
  });
});

describe('postData Saga', () => {
  const postDataSaga = postData();

  it('should start task to watch for LOAD_POST action', () => {
    const takeLatestDescriptor = postDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_POST, getPost));
  });
});
