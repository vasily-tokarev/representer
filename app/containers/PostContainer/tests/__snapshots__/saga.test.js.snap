// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPost Saga should dispatch the getPost action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getPost Saga should dispatch the getPost action 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/posts/name/[object Object].json",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
